### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Zelinskyi-Serhii**
[DEMO LINK](https://Zelinskyi-Serhii.github.io/react_pagination/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -33,64 +48,13 @@ export const App: React.FC = () => {
...
-      </ul>
-      <ul>
-        <li data-cy="item">Item 1</li>
-        <li data-cy="item">Item 2</li>
-        <li data-cy="item">Item 3</li>
-        <li data-cy="item">Item 4</li>
-        <li data-cy="item">Item 5</li>
-      </ul>
+      <Pagination
+        total={42}
```
Comment (KuharchukDmytro) [inline src/App.tsx:52]:
```suggestion
        total={items.length}
```

File: src/App.tsx [hunk 2]
```diff
@@ -21,6 +31,11 @@ export const App: React.FC = () => {
             data-cy="perPageSelector"
             id="perPageSelector"
             className="form-control"
+            value={itemsPerPage}
+            onChange={({ target }) => {
+              setPerPage(+target.value);
+              setCurrentPage(1);
+            }}
```
Comment (KuharchukDmytro) [inline src/App.tsx:38]: better to make separate handler for this changeEvent

File: src/App.tsx [hunk 3]
```diff
@@ -33,64 +48,13 @@ export const App: React.FC = () => {
...
-        <li data-cy="item">Item 3</li>
-        <li data-cy="item">Item 4</li>
-        <li data-cy="item">Item 5</li>
-      </ul>
+      <Pagination
+        total={42}
+        perPage={itemsPerPage}
+        currentPage={currentPage}
+        onPageChange={setCurrentPage}
+        visibleItems={visibleItems}
```
Comment (KuharchukDmytro) [inline src/App.tsx:56]: rely on task condition, there must be only four props
image

File: src/components/Pagination/Pagination.tsx [hunk 4]
```diff
@@ -1 +1,100 @@
...
+            if (currentPage !== i) {
+              onPageChange(i);
+            }
+          }}
+        >
+          {i}
+        </a>
+      </li>,
+    );
+  }
```
Comment (KuharchukDmytro) [inline src/components/Pagination/Pagination.tsx:49]: you have numbers of pages, on my opinion you can just use helper getNumbers function to make an array of pages count, and then just map through it

File: src/components/Pagination/Pagination.tsx [hunk 5]
```diff
@@ -1 +1,100 @@
...
+      <ul>
+        {visibleItems.map(item => (
+          <li
+            data-cy="item"
+            key={uuidv4()}
+          >
+            {item}
+          </li>
+        ))}
+      </ul>
```
Comment (KuharchukDmytro) [inline src/components/Pagination/Pagination.tsx:97]: it should be in app file, in this file you only render pagination

File: src/components/Pagination/Pagination.tsx [hunk 6]
```diff
@@ -1 +1,100 @@
...
+              }
+            }}
+          >
+            «
+          </a>
+        </li>
+
+        {listItem}
+
+        <li className={classNames('page-item', { disabled: isLastPage })}>
```
Comment (KuharchukDmytro) [inline src/components/Pagination/Pagination.tsx:72]:
```suggestion
        <li className={classNames(
            'page-item',
            {
              disabled: isLastPage
            },
          )}
         >
```

**Review by KuharchukDmytro — CHANGES_REQUESTED** [review]
good job, but there are several mistakes you need to fix👨‍💻

File: src/App.tsx [hunk 7]
```diff
@@ -1,18 +1,34 @@
-import React from 'react';
+import React, { useState } from 'react';
+import { v4 as uuidv4 } from 'uuid';
 import './App.css';
 import { getNumbers } from './utils';
+import { Pagination } from './components/Pagination/Pagination';
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
 const items = getNumbers(1, 42)
```
Comment (sergii-nosachenko) [inline src/App.tsx:7]: Create `constants.ts` and add 42 as constant there.

File: src/App.tsx [hunk 8]
```diff
@@ -1,18 +1,34 @@
...
   .map(n => `Item ${n}`);
 
 export const App: React.FC = () => {
+  const [itemsPerPage, setPerPage] = useState<number>(5);
```
Comment (sergii-nosachenko) [inline src/App.tsx:11]: You don't need to specify the type, it will be `number` by default (as typeof 5 is number)

Comment (sergii-nosachenko) [inline src/App.tsx:11]: 5 --> constants

File: src/App.tsx [hunk 9]
```diff
@@ -1,18 +1,34 @@
...
+  const [currentPage, setCurrentPage] = useState(1);
+
+  const howManyVisibleItems = itemsPerPage * currentPage - itemsPerPage + 1;
```
Comment (sergii-nosachenko) [inline src/App.tsx:14]:
```suggestion
  const visibleItemsCount = itemsPerPage * currentPage - itemsPerPage + 1;
```

File: src/App.tsx [hunk 10]
```diff
@@ -1,18 +1,34 @@
...
+    setCurrentPage(1);
+  };
+
   return (
     <div className="container">
       <h1>Items with Pagination</h1>
 
       <p className="lead" data-cy="info">
-        Page 1 (items 1 - 5 of 42)
+        {`Page ${currentPage} (items ${howManyVisibleItems} - ${lastVisibleItem} of 42)`}
```
Comment (sergii-nosachenko) [inline src/App.tsx:31]: Get42 from constants

File: src/App.tsx [hunk 11]
```diff
@@ -21,6 +37,8 @@ export const App: React.FC = () => {
             data-cy="perPageSelector"
             id="perPageSelector"
             className="form-control"
+            value={itemsPerPage}
+            onChange={handleSetPerPage}
           >
             <option value="3">3</option>
```
Comment (sergii-nosachenko) [inline src/App.tsx:43]: Generate options from array

File: src/App.tsx [hunk 12]
```diff
@@ -33,63 +51,22 @@ export const App: React.FC = () => {
           items per page
         </label>
       </div>
+      <Pagination
```
Comment (sergii-nosachenko) [inline src/App.tsx:54]: Add space

File: src/App.tsx [hunk 13]
```diff
@@ -33,63 +51,22 @@ export const App: React.FC = () => {
...
-            data-cy="nextLink"
-            className="page-link"
-            href="#next"
-            aria-disabled="false"
-          >
-            »
-          </a>
-        </li>
-      </ul>
       <ul>
```
Comment (sergii-nosachenko) [inline src/App.tsx:61]: Add space

File: src/App.tsx [hunk 14]
```diff
@@ -33,63 +51,22 @@ export const App: React.FC = () => {
...
-        <li data-cy="item">Item 1</li>
-        <li data-cy="item">Item 2</li>
-        <li data-cy="item">Item 3</li>
-        <li data-cy="item">Item 4</li>
-        <li data-cy="item">Item 5</li>
+        {visibleItems.map(item => (
+          <li
+            data-cy="item"
+            key={uuidv4()}
```
Comment (sergii-nosachenko) [inline src/App.tsx:65]:
```suggestion
            key={item}
```
In this task item can be used as a key, but your solution is ok

File: src/components/Pagination/Pagination.tsx [hunk 15]
```diff
@@ -1 +1,89 @@
...
+};
+
+export const Pagination: React.FC<Props> = ({
+  perPage,
+  currentPage,
+  onPageChange,
+  total,
+}) => {
+  const isFirstPage = currentPage === 1;
+  const numbersOfPage = Math.ceil(total / perPage);
```
Comment (sergii-nosachenko) [inline src/components/Pagination/Pagination.tsx:19]:
```suggestion
  const pagesCount = Math.ceil(total / perPage);
```

File: src/components/Pagination/Pagination.tsx [hunk 16]
```diff
@@ -1 +1,89 @@
...
+  const isLastPage = currentPage === numbersOfPage;
+
+  const listItem = getNumbers(1, numbersOfPage)
```
Comment (sergii-nosachenko) [inline src/components/Pagination/Pagination.tsx:22]:
```suggestion
  const pagesListItems = getNumbers(1, numbersOfPage)
```

File: src/components/Pagination/Pagination.tsx [hunk 17]
```diff
@@ -1 +1,89 @@
...
+          { disabled: isLastPage },
+        )}
+        >
+          <a
+            data-cy="nextLink"
+            className="page-link"
+            href="#next"
+            aria-disabled={isLastPage}
+            onClick={() => {
+              if (currentPage !== numbersOfPage) {
```
Comment (sergii-nosachenko) [inline src/components/Pagination/Pagination.tsx:78]:
```suggestion
              if (!isLastPage) {
```

File: src/components/Pagination/Pagination.tsx [hunk 18]
```diff
@@ -1 +1,89 @@
...
+    <>
+      <ul className="pagination">
+        <li className={classNames('page-item', { disabled: isFirstPage })}>
+          <a
+            data-cy="prevLink"
+            className="page-link"
+            href="#prev"
+            aria-disabled={isFirstPage}
+            onClick={() => {
+              if (currentPage !== 1) {
```
Comment (sergii-nosachenko) [inline src/components/Pagination/Pagination.tsx:56]: Use `isFirstPage` here

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 19]
```diff
@@ -40,19 +42,24 @@ export const App: React.FC = () => {
             value={itemsPerPage}
             onChange={handleSetPerPage}
           >
-            <option value="3">3</option>
-            <option value="5">5</option>
-            <option value="10">10</option>
-            <option value="20">20</option>
+            {Object.values(ItemsPerPage).map(item => (
```
Comment (sergii-nosachenko) [inline src/App.tsx:45]: A simple array will be better

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]

File: src/constants.ts [hunk 20]
```diff
@@ -1 +1,7 @@
+import { getNumbers } from './utils';
+
 export const totalItems = 42;
+export const itemsPerPageConstant = [3, 5, 10, 20];
```
Comment (sergii-nosachenko) [inline src/constants.ts:4]:
```suggestion
export const itemsPerPageOptions = [3, 5, 10, 20];
export const defaultItemsPerPage = itemsPerPageOptions[1];
```

File: src/App.tsx [hunk 21]
```diff
@@ -2,26 +2,21 @@ import React, { useState } from 'react';
...
-import { ItemsPerPage } from './types.ts/ItemsPerPage';
-import { totalItems } from './constants';
-
-const items = getNumbers(1, totalItems)
-  .map(n => `Item ${n}`);
+import { items, itemsPerPageConstant, totalItems } from './constants';
 
 export const App: React.FC = () => {
-  const [itemsPerPage, setPerPage] = useState<ItemsPerPage>(ItemsPerPage.FIVE);
+  const [itemsPerPage, setPerPage] = useState(5);
```
Comment (sergii-nosachenko) [inline src/App.tsx:8]:
```suggestion
  const [itemsPerPage, setPerPage] = useState(defaultItemsPerPage);
```

**Review by sergii-nosachenko — APPROVED** [review]

**Pull request closed without merging**